src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingWriteStream.cs

```diff
@@ -0,0 +1,204 @@
+        {
+            ThrowArgumentException(buffer, offset, count);
+            ArraySegment<byte> bufferSegment = new ArraySegment<byte>(buffer, offset, count);
+            return WriteAsync(bufferSegment, cancellationToken);
+        }
+
+        private async Task WriteAsync(
+            ArraySegment<byte> bufferSegment,
+            CancellationToken cancellationToken)
+        {
+            bool decoderCompleted = false;
+
+            while (!decoderCompleted)
+            {
+#if NETSTANDARD2_0
+                _decoder.Convert(bufferSegment.Array, bufferSegment.Offset, bufferSegment.Count, _charBuffer, _charsDecoded, _charBuffer.Length - _charsDecoded,
+                    flush: false, out int bytesDecoded, out int charsDecoded, out decoderCompleted);
+#else
+                _decoder.Convert(
+                    bufferSegment,
+                    _charBuffer.AsSpan(_charsDecoded),
+                    flush: false,
+                    out int bytesDecoded,
+                    out int charsDecoded,
+                    out decoderCompleted);
+#endif
+
+                _charsDecoded += charsDecoded;
+                bufferSegment = bufferSegment.Slice(bytesDecoded);
+
+                if (!decoderCompleted)
+                {
+                    await WriteBufferAsync(cancellationToken).ConfigureAwait(false);
+                }
+            }
+        }
+
+        private async Task WriteBufferAsync(CancellationToken cancellationToken)
+        {
+            bool encoderCompleted = false;
+            int charsWritten = 0;
+            byte[] byteBuffer = ArrayPool<byte>.Shared.Rent(_maxByteBufferSize);
+
+            while (!encoderCompleted && charsWritten < _charsDecoded)
+            {
+#if NETSTANDARD2_0
+                _encoder.Convert(_charBuffer, charsWritten, _charsDecoded - charsWritten, byteBuffer, 0, byteBuffer.Length,
+                    flush: false, out int charsEncoded, out int bytesUsed, out encoderCompleted);
+
+                await _stream.WriteAsync(byteBuffer, 0, bytesUsed, cancellationToken).ConfigureAwait(false);
+#else
+                _encoder.Convert(
+                    _charBuffer.AsSpan(charsWritten, _charsDecoded - charsWritten),
+                    byteBuffer,
+                    flush: false,
+                    out int charsEncoded,
+                    out int bytesUsed,
+                    out encoderCompleted);
+
+                await _stream.WriteAsync(byteBuffer.AsMemory(0, bytesUsed), cancellationToken).ConfigureAwait(false);
+#endif
+
+                charsWritten += charsEncoded;
+            }
+
+            ArrayPool<byte>.Shared.Return(byteBuffer);
+
+            // At this point, we've written all the buffered chars to the underlying Stream.
+            _charsDecoded = 0;
+        }
+
+        private static void ThrowArgumentException(byte[] buffer, int offset, int count)
+        {
```

Comment:
Same comments as on the read stream: validation of buffer being non-null, method name, etc.